### ISSUE TITLE
Assertion m_private_counter >= 0 failed (located in the destroy function, line in file: 141)

### ISSUE THREAD

**Issue by xiamr** [oneapi-src/oneTBB#639]
When using flow graph, the following assert failed with the following stderr output:

**Assertion m_private_counter >= 0 failed (located in the destroy function, line in file: 141)
Detailed description: Private counter may not be less than 0**

in the destroy function:
![1636081508(1)](https://user-images.githubusercontent.com/18632312/140451595-403f268e-dd0a-485d-9920-edc7a14f3f63.png)

Call Stack:
![image](https://user-images.githubusercontent.com/18632312/140452619-feae5168-8c99-428f-b573-c0d31e9be659.png)

_raise 0x000014a314b6637f
abort 0x000014a314b50db5
tbb::detail::r1::assertion_failure_impl assert_impl.h:56
operator() assert_impl.h:73
tbb::detail::d0::run_initializer<tbb::detail::r1::assertion_failure(char const*, int, char const*, char const*)::<lambda()> >(const struct {...} &, std::atomic<tbb::detail::d0::do_once_state> &) _utils.h:288
tbb::detail::d0::atomic_do_once<tbb::detail::r1::assertion_failure(char const*, int, char const*, char const*)::<lambda()> >(const struct {...} &, std::atomic<tbb::detail::d0::do_once_state> &) _utils.h:277
tbb::detail::r1::assertion_failure assert_impl.h:73
tbb::detail::r1::small_object_pool_impl::destroy small_object_pool.cpp:141
tbb::detail::r1::thread_data::~thread_data thread_data.h:120
tbb::detail::r1::governor::auto_terminate governor.cpp:221
tbb::detail::r1::market::cleanup market.cpp:611
tbb::detail::r1::rml::private_worker::run private_server.cpp:283
tbb::detail::r1::rml::private_worker::thread_routine private_server.cpp:221
start_thread 0x000014a3156a815a
clone 0x000014a314c2bdd3_

**Comment by xiamr** [comment]
the issue is existed in oneTBB 2021.3 and 2021.4 version

**Comment by alexey-katranov** [comment]
The issue seems weird like double-free or something like that. Can you provide mode details about your reproducer? What flow graph nodes are used and how? Is it x86?

**Comment by xiamr** [comment]
It is x86-64. And the code has this issue：
![企业微信截图_16364373962284](https://user-images.githubusercontent.com/18632312/140870342-12728f6d-6990-4999-b9cb-c327ef256440.png)

**Comment by pranasge** [comment]
It seems I have this issue to and it can be reproduced with simple test:
```	
	struct TestStruct
	{
		char bytes[153]{ 0 };
		//char bytes[152]{ 0 }; //works
	};

	tbb::flow::graph graph;
	tbb::flow::queue_node<TestStruct> input{ graph };
	tbb::flow::function_node<TestStruct> func{ graph, tbb::flow::serial, [](const auto & input) { return input; } };

	tbb::flow::make_edge(input, func);

	ASSERT_TRUE(input.try_put(TestStruct{}));

	graph.wait_for_all();
```
Tested with oneTBB 2021.3 version.

**Comment by pranasge** [comment]
This  seems to be blocking  issue for me. Is threre easy workaround or version that does not have this issue?

**Comment by alexey-katranov** [comment]
We are in the process of fixing. I am not sure about possible workaround (some crazy variants: reduce the message size, e.g. use the shared pointer or recompile oneTBB runtime with the disabled small object cache).
The root cause is that we detect the object size incorrectly during destruction of a graph message, overall, it is unlikely to introduce any issues in release (while there is possible race condition in a corner case if threads are heavily recreated during graph processing)

**Comment by pranasge** [comment]
Wrong deallocation sounded bad but seems to be less of an issue than I thought. Will use release version for now. Thanks for info.

**Comment by alexey-katranov** [comment]
@pranasge , do you have the possibility to check #667 that it fixes your original problem?

**Comment by pranasge** [comment]
Yes, my tests pass without problems now.

**Comment by Veronizlada** [comment]
I am having same issue. Is this in prod? Or I have to check out a specific commit?

**Comment by vlserov** [comment]
Commit [6edb5c3](https://github.com/oneapi-src/oneTBB/commit/6edb5c3ae98c86cd42453b7241dcbcba49dc80c8) was not became part of [v2021.5.0](https://github.com/oneapi-src/oneTBB/releases/tag/v2021.5.0).
Should be included in latest [v2021.6.0](https://github.com/oneapi-src/oneTBB/releases/tag/v2021.6.0) release.

**Comment by Nekto89** [comment]
google has led me here. I'm getting this assertion non-deterministically on v2021.9.0 on Linux and macOS. I haven't been able to minimize the case yet.
